File: python3/Graphs/graph_deep_copy.py
from ds import GraphNode

"""
Definition of GraphNode:
class GraphNode:
    def __init__(self, val):
        self.val = val
        self.neighbors = []
"""


def graph_deep_copy(node: GraphNode) -> GraphNode:
    if not node:
        return None
    return dfs(node)

def dfs(node: GraphNode, clone_map = {}) -> GraphNode:
    # If this node was already cloned, then return this previously  
    # cloned node.
    if node in clone_map:
        return clone_map[node]
    # Clone the current node.
    cloned_node = GraphNode(node.val)
    # Store the current clone to ensure it doesn't need to be created 
    # again in future DFS calls.
    clone_map[node] = cloned_node
    # Iterate through the neighbors of the current node to connect  
    # their clones to the current cloned node.
    for neighbor in node.neighbors:
        cloned_neighbor = dfs(neighbor, clone_map)
        cloned_node.neighbors.append(cloned_neighbor)
    return cloned_node

File: cpp/Graphs/graph_deep_copy.cpp
#include <unordered_map>
#include <vector>
#include "ds/GraphNode.h"
using ds::GraphNode;

/**
 * Definition of GraphNode:
 * class GraphNode {
 * public:
 *     int val;
 *     std::vector<GraphNode*> neighbors;
 *     GraphNode(int _val) {
 *         val = _val;
 *         neighbors = vector<GraphNode*>();
 *     }
 * };
 */

GraphNode* graphDeepCopy(GraphNode* node) {
    if (!node) {
        return nullptr;
    }
    std::unordered_map<GraphNode*, GraphNode*> cloneMap;
    return dfs(node, cloneMap);
}

GraphNode* dfs(GraphNode* node, std::unordered_map<GraphNode*, GraphNode*>& cloneMap) {
    // If this node was already cloned, then return this previously
    // cloned node.
    if (cloneMap.find(node) != cloneMap.end()) {
        return cloneMap[node];
    }
    // Clone the current node.
    GraphNode* clonedNode = new GraphNode(node->val);
    // Store the current clone to ensure it doesn't need to be created
    // again in future DFS calls.
    cloneMap[node] = clonedNode;
    // Iterate through the neighbors of the current node to connect
    // their clones to the current cloned node.
    for (GraphNode* neighbor : node->neighbors) {
        GraphNode* clonedNeighbor = dfs(neighbor, cloneMap);
        clonedNode->neighbors.push_back(clonedNeighbor);
    }
    return clonedNode;
}